### DISCUSSION TITLE
Infinite loop in curl_multi_perform when response time is too high from server.

### DISCUSSION THREAD

**Discussion by bansalnehal** [curl/curl#12642] [Q&A] [answered]
Hi, I have an implementation as following: 
There are two pthreads. The easy handles containing the requests are added using curl_multi_add_handle on one thread (Thread 1). And curl_multi_perform with curl_multi_info_read are executed on a different thread (Thread 2). Both the threads are using the pthread_mutex locking mechanism such that there is no concurrent access to the multi handle. It is mostly working fine but sometimes there occurs a situation where the responses are delayed from the server. At that moment, Thread 1 which is supposed to add handles is waiting to acquire the mutex lock, and Thread 2 executing curl_multi_perform is stuck in a loop between Curl_splaygetbest and add_next_timeout functions and is functioning at 100% CPU. How should I proceed with this? Because after this the entire process is stuck.

**Comment by bagder** [comment]
It is either a bad API use by you or a libcurl issue. It is hard to tell which unless you can provide a recipe that reproduces the problem (with a modern libcurl version).

**Reply by bansalnehal** [reply]
I am using curl 7.76.1.
Following is the pseudo code:

Thread 1:

Waits on a queue from which HTTP request is to be made from.
Make an easy handle for the request.
Acquire mutex lock and add the easy handle to the multi handle.
Relinquish the lock.

Thread 2:

curl_multi_wait is called on the multi_handle.
mutex lock is acquired
curl_multi_perform and curl_multi_info_read is called.
mutex lock is released.

**Comment by jay** [comment]
> The easy handles containing the requests are added using curl_multi_add_handle on one thread (Thread 1). And curl_multi_perform with curl_multi_info_read are executed on a different thread (Thread 2). Both the threads are using the pthread_mutex locking mechanism such that there is no concurrent access to the multi handle.

Internal splay, hash, linked list busy loops or crashes in curl are often the sign of bad multithreading, like you forgot a lock somewhere. Please review [thread safety guidelines](https://curl.se/libcurl/c/threadsafe.html).

Or it could be a bug. Try the latest curl. If you need help after reviewing the guidelines and trying the latest curl please include a _minimal_ self contained example that can be used to reproduce. If you really can't do that then paste in relevant parts of the code (like make some pseudo code) and I'll try to determine if there's anything wrong with the way you are doing it.

Depending on what you are doing, for performance a better way may be no multi lock, make your own thread safe queue of easy handles, in Thread 1 take ownership of that queue (eg empty it), release the lock, call curl_multi_add_handle on those queued handles.

**Reply by bansalnehal** [reply]
I am using curl 7.76.1. 
Following is the pseudo code: 
Thread 1:
1) Waits on a queue from which HTTP request is to be made from. 
2) Make an easy handle for the request. 
3) Acquire mutex lock and add the easy handle to the multi handle. 
4) Relinquish the lock. 

Thread 2: 
1) curl_multi_wait is called on the multi_handle. 
2) mutex lock is acquired
3) curl_multi_perform and curl_multi_info_read is called. 
4) mutex lock is released.

**Reply by jay** [reply]
If you do it that way all calls operating on the multi have to be locked. It is not very efficient and will back up your worker threads. If you are calling curl_multi_wait without the lock that's a problem.

**Reply by bansalnehal** [reply]
But how do I get out of this situation? I tried removing the multi_wait and instead made Thread 2 sleep for 1 usec. That increased my CPU utilization. As there happens a situation where curl_multi_perform is working at 100%  at some __sigaction as seen in gdb.

**Reply by bansalnehal** [reply]
I have another query. What does curl do of requests if HTTP server is not reachable? Because I have seen that this Thread 2 then goes at 100%.

**Reply by jay** [reply]
You should call curl_multi_wait when you have the lock. Though as I said earlier the way you are doing it is not efficient and a better way may be no multi lock and you lock a queue instead.

> What does curl do of requests if HTTP server is not reachable? Because I have seen that this Thread 2 then goes at 100%.

First solve the multithreading problems, since what you're seeing may be a symptom of that.

**Reply by bansalnehal** [reply]
I have removed the call of multi_wait without the lock. 

The new implementation looks like this: 
Thread 1:

1) Waits on a queue from which HTTP request is to be made from.
2) Make an easy handle for the request.
3) Acquire mutex lock and add the easy handle to the multi handle.
4) Relinquish the lock.

Thread 2:

1) sleep for 1 usec
2) mutex lock is acquired
3) curl_multi_perform and curl_multi_info_read is called.
4) mutex lock is released.

This is also yielding 100% CPU inside multi_perform somewhere.

**Reply by bansalnehal** [reply]
> "You should call curl_multi_wait when you have the lock. Though as I said earlier the way you are doing it is not efficient and a better way may be no multi lock and you lock a queue instead." 

Which queue are you talking about in this? Also what is multi lock? 

Also, my target is to keep the thread at an optimum CPU usage even if I don't get responses, since that is to be handled at the server. Even if the server is not responding or unreachable or delayed, I don't want my thread running at 100%. Is there a way to do that?

**Reply by jay** [reply]
Call curl_multi_wait (or curl_multi_poll). Why did you remove the call?

> Which queue are you talking about in this? Also what is multi lock?

I explained [earlier](https://github.com/curl/curl/discussions/12642#discussioncomment-8033892) that for performance it is better not to lock when using the multi. What I meant by "multi lock" is that: the mutex you are using to lock every time you use the multi.

**Reply by bansalnehal** [reply]
If I use multi wait on Thread 2, I would have had to use it inside lock, and till then Thread 1 would have waited to add on more easy handles. For reducing wait for Thread 1 I removed multi wait. Also I was suspecting that the 100% cpu after server stops responding could be because of multi wait being used outside the lock. 

If I remove the lock, how would I use the two threads. 
I can’t lock the queue, since the requests which are to be made into easy handles are coming from a different interface from where regular messages would be received.

**Reply by jay** [reply]
> If I use multi wait on Thread 2, I would have had to use it inside lock, and till then Thread 1 would have waited to add on more easy handles.

Exactly right. That's why it's bad for performance to do it that way.

> If I remove the lock, how would I use the two threads.

I explained one way you could do it by making your own thread safe queue of easy handles and then use the multi only from Thread 2.

**Reply by bansalnehal** [reply]
Okay. I’ll try this and get back.

**Reply by bansalnehal** [reply]
I tried this. The thread is still at 100% cpu. Following is the pseudo code of Thread 2: 
while(true)
  1) Fetch all easy handles that were created by Thread 1. 
  2) do 
      a. curl_multi_wait
      b. curl_multi_perform
      c. while (msg = curl_multi_info_read)
          c.i As soon as a message is done, add another easy handle created by Thread 1 to the multi handle. 
      while(still alive && msg)

It is stuck somewhere is Curl_now() inside curl_multi_perform. This time there is no actual message exchange going on. The system is idle. 


I have two more queries: 
1) Is there a callback function to find out if the connection has timed out for a easy handle added to a multi handle. I tried CURLOPT_TIMEOUT, but don't know what to do further with it. 
2) If a request times out, what happens to the request's easy handle which was part of the multi handle?

**Reply by bagder** [reply]
> If a request times out, what happens to the request's easy handle which was part of the multi handle?

That transfer ends with a `CURLE_OPERATION_TIMEDOUT` error code.

**Reply by bansalnehal** [reply]
How do I detect this error on a multi handle?

**Reply by bagder** [reply]
It is a completed transfer. curl_multi_info_read tells you about completed transfers.

**Reply by bansalnehal** [reply]
> That transfer ends with a CURLE_OPERATION_TIMEDOUT error code.

This worked. Thank you!


The idle CPU usage got resolved with curl_multi_poll, even when unreachable along with continuous transfer, it is at 3%. 

But it is still at 100 when continuous transfer is being tried but the server is slow in responding. Could you suggest some solution?

**Reply by jay** [reply]
> The idle CPU usage got resolved with curl_multi_poll, even when unreachable along with continuous transfer, it is at 3%.
> 
> But it is still at 100 when continuous transfer is being tried but the server is slow in responding. Could you suggest some solution?

There's either a bug in your program or a bug in curl. If you can reproduce using the latest curl please make a minimal self contained example and we'll try it.

**Reply by bansalnehal** [reply]
Actually, due to some dependencies I can't upgrade the curl version. The code for Thread 2 is as follows: 

~~~c
CURLMsg *msg; 
int msgs_left = -1; 
int still_alive = 0; 
do 
{
        //read all and empty the easy handle queue and add all handles using curl_multi_add_handle. (thread-safe)
        int numfds = 0; 
   	do 
      	{ 
   		CURLMcode res = curl_multi_poll(multi ,NULL,0,1000, &numfds);  
    		curl_multi_perform(multi, &still_alive); 
	    while ((msg = curl_multi_info_read(multi, &msgs_left))) 
            { 
                if (msg->msg == CURLMSG_DONE) 
		{ 
                    char *url; 
                    CURL *e = msg->easy_handle; 
                    
                    curl_easy_getinfo(msg->easy_handle, CURLINFO_PRIVATE, &url);                     
                    curl_multi_remove_handle(multi, e); 
                    curl_easy_cleanup(e);

                    //read one handle from easy handle queue and add using curl_multi_add_handle.  (thread-safe)
                } 
            } 
        } while(still_alive && msg); 
} while (1);
~~~

**Reply by bansalnehal** [reply]
Hi, could you find the issue with my code?

**Reply by jay** [reply]
> Hi, could you find the issue with my code?

> ```c
>         } while(still_alive && msg); 
> } while (1);
> ```

You have a loop that will break if there are no running handles (`still_alive` is 0), or there were no messages available for any of the running handles (`msg` is NULL). That is wrong. There may not be any messages. If there are no running handles (as in no transfers in progress, and in your code all handles removed) then you would wait until the thread queue has more handles.

~~You have an inifinite outer loop `while(1)` that will eat cpu if there are no handles in the multi.  That is wrong. I'm sure this is pseudo code so it's more complicated than that, but if not that could be why it's eating CPU. I would debug that to find out.~~

**Reply by bansalnehal** [reply]
Hi, according to me, it goes like this: 
It checks if there is anything in the easy handle queue, 
If nothing, it lands on multi poll. 
The loop on while(1) is doing the above 2 steps. 

Also, the 100% that I see is when there are continuous requests coming. It works seamlessly when the server responds in time. The 100% occurs when the server is faulty.

**Reply by jay** [reply]
> If nothing, it lands on multi poll.

Ah I missed that. Yes it should wait the 1 second if there are no running handles. Using two loops like that is a weird way to do it. You would have to debug to figure out what is eating cpu.

**Reply by bansalnehal** [reply]
Actually it is a pthread. The outer while is just to see if any new requests are received. If I don’t use it, the thread would just exit once all the transfers are completed. So, when my next set of messages are received, nothing would happen.

**Reply by jay** [reply]
Ok. All I can tell you at this point is use the latest curl and if you can reproduce then give us a minimal self contained example. Without those two things there's nothing else we can do for you here.

**Reply by bansalnehal** [reply]
Okay. Thank you for inputs, they've helped a lot. Due to certain constraints, I am unable to upgrade curl. There are just two more questions: 

1) Does curl_multi_poll exit when there is a message to be received from the server, and then multi_perform is to be called? 
2) There is a certain point in my testcase when the multi_perform is doing the below function calls (CPU reaching 100) and the number of connections have increased from 90 to 900.  I used gdb on the running thread. Could you tell the reason. 
 a) 
No symbol table info available.
 0x00000039365f104e in X509_STORE_load_locations ()
   from /usr/lib64/libcrypto.so.1.1
No symbol table info available.
 0x00007f564d962cbb in ossl_connect_step1 () from /usr/lib64/libcurl.so.4
No symbol table info available.
 0x00007f564d96337e in ossl_connect_common () from /usr/lib64/libcurl.so.4
No symbol table info available.
 0x00007f564d964f72 in Curl_ssl_connect_nonblocking ()
   from /usr/lib64/libcurl.so.4
No symbol table info available.
 0x00007f564d9276d6 in https_connecting () from /usr/lib64/libcurl.so.4
No symbol table info available.
0x00007f564d92a67b in Curl_http_connect () from /usr/lib64/libcurl.so.4
No symbol table info available.
 0x00007f564d938d95 in multi_runsingle () from /usr/lib64/libcurl.so.4
No symbol table info available.
 0x00007f564d939b6b in curl_multi_perform () from /usr/lib64/libcurl.so.4
No symbol table info available.
 
b)
  0x00000037c280f8fe in __libc_sigaction () from /lib64/libpthread.so.0
No symbol table info available.
  0x00007efcbfd77adf in sigpipe_ignore () from /usr/lib64/libcurl.so.4
No symbol table info available.
  0x00007efcbfd79b5b in curl_multi_perform () from /usr/lib64/libcurl.so.4
No symbol table info available.

The (b) case above is making it get stuck at 100% and then the process restarts.

**Reply by jay** [reply]
> 1. Does curl_multi_poll exit when there is a message to be received from the server, and then multi_perform is to be called?

Yes, [curl_multi_poll](https://curl.se/libcurl/c/curl_multi_poll.html) will return when activity is detected.

> 2. There is a certain point in my testcase when the multi_perform is doing the below function calls (CPU reaching 100) and the number of connections have increased from 90 to 900.  I used gdb on the running thread. Could you tell the reason.
> 
> No symbol table info available. #13 0x00000039365f104e in X509_STORE_load_locations ()

That's a lot of connections, maybe it has something to do with the high CPU usage. Do you expect that many connections?

If OpenSSL is trying to parse and load a certificate bundle for many connections one after another repeatedly then I guess it's possible it could eat a lot of cpu doing that. If I were you I'd profile the application to see where the CPU usage is going, using something like perf.

If it is the problem:

curl 7.87.0 and later [supports a CA cache](https://daniel.haxx.se/blog/2022/11/11/curls-new-ca-store-cache/) of the last loaded certificate bundle in a multi. If all easy handles in the multi use the same certificate bundle that is the ideal situation because then one bundle stays cached and doesn't have to be read again.

Since you are using an older version the options aren't as good. If your curl is using the certificates maintained by your OS and both the default CA file and CA path (you would see both in verbose output as `*  CAfile: xxx` and `*  CApath: xxx` and not `* CApath: none`) then you could try disabling CA file in each easy handle with `curl_easy_setopt(curl, CURLOPT_CAINFO, NULL);` and OpenSSL would just use the default CApath which _might_ be faster. Another way would be to implement your own CA caching of the certificate bundle by using CURLOPT_SSL_CTX_FUNCTION to manage certificates yourself. (I'm not sure if this is a good idea or not in old versions)

**Reply by bansalnehal** [reply]
I found the problem. The CApath was none when I checked with verbose. 
I had set the CONNECTTIMEOUT parameter only, not the TIMEOUT. So, if the connection got established, but the server is lagging, the easy handle would not timeout until 300 seconds, the default value of TIMEOUT. Hence, curl was increasing connections to keep up with new added handles because I had also not put an upper limit on the number of connections. I reduced the timeout to 2 seconds, so now even if the thread is reaching 98-99 which is due to the faulty server, the process doesn't restart, as the handles are getting cleared on timeout. 

The discussion has been of huge help. Thank you @jay and @bagder . Please wait for a while before closing the discussion in case I need to ask further queries. Thanks!

**Reply by jay** [reply]
To be clear [CURLOPT_CONNECTTIMEOUT](https://curl.se/libcurl/c/CURLOPT_CONNECTTIMEOUT.html) default value is 300 seconds (300) and [CURLOPT_TIMEOUT](https://curl.se/libcurl/c/CURLOPT_TIMEOUT.html) default value is no timeout (0). See also [CURLMOPT_MAX_TOTAL_CONNECTIONS](https://curl.se/libcurl/c/CURLMOPT_MAX_TOTAL_CONNECTIONS.html) and [CURLMOPT_MAX_HOST_CONNECTIONS](https://curl.se/libcurl/c/CURLMOPT_MAX_HOST_CONNECTIONS.html), both of which have no limit (0) as the default value.

In my opinion, if something is still eating CPU then you have not really solved the problem.

**Reply by bansalnehal** [reply]
This is the gdb output of the thread executing multi_perform - 
#0  0x0000003936594684 in PEM_read_bio_ex () from /usr/lib64/libcrypto.so.1.1
No symbol table info available.
#1  0x00000039365928e4 in PEM_X509_INFO_read_bio () from /usr/lib64/libcrypto.so.1.1
No symbol table info available.
#2  0x00000039365edf01 in X509_load_cert_crl_file () from /usr/lib64/libcrypto.so.1.1
No symbol table info available.
#3  0x00000039365ee070 in by_file_ctrl () from /usr/lib64/libcrypto.so.1.1
No symbol table info available.
#4  0x00000039365f104e in X509_STORE_load_locations () from /usr/lib64/libcrypto.so.1.1
No symbol table info available.
#5  0x00007f0f6d251cbb in ossl_connect_step1 () from /usr/lib64/libcurl.so.4
No symbol table info available.
#6  0x00007f0f6d25237e in ossl_connect_common () from /usr/lib64/libcurl.so.4
No symbol table info available.
#7  0x00007f0f6d253f72 in Curl_ssl_connect_nonblocking () from /usr/lib64/libcurl.so.4
No symbol table info available.
#8  0x00007f0f6d2166d6 in https_connecting () from /usr/lib64/libcurl.so.4
No symbol table info available.
#9  0x00007f0f6d21967b in Curl_http_connect () from /usr/lib64/libcurl.so.4
No symbol table info available.
#10 0x00007f0f6d227d95 in multi_runsingle () from /usr/lib64/libcurl.so.4
No symbol table info available.
#11 0x00007f0f6d228b6b in curl_multi_perform () from /usr/lib64/libcurl.so.4
No symbol table info available.

I have checked multiple times, this was the output each time the CPU was closing in on 100. 

And yes, I did not use CURLMOPT_MAX_TOTAL_CONNECTIONS and CURLMOPT_MAX_HOST_CONNECTIONS, because in case of delay on one message, I did not want to delay the transfer of another, which would have taken place on the same set of connections (if the limit has been set). Is my understanding correct? 

Another question, is it necessary to complete all the transfers, before adding any new easy handles to the multi? I modified my implementation as below:  

  ```
CURLMsg *msg; 
   int msgs_left = -1; 
   int still_alive = 0; 
   do 
   {
         //read all and empty the easy handle queue and add all handles using **curl_multi_add_handle.** (thread-safe)
         int numfds = 0; 
         CURLMcode res = curl_multi_poll(multi ,NULL,0,1000, &numfds);  
         curl_multi_perform(multi, &still_alive); 
         while ((msg = curl_multi_info_read(multi, &msgs_left))) 
         { 
               if (msg->msg == CURLMSG_DONE) 
               { 
                    char *url; 
                    CURL *e = msg->easy_handle; 
                    curl_easy_getinfo(msg->easy_handle, CURLINFO_PRIVATE, &url);                     
                    curl_multi_remove_handle(multi, e); 
                    curl_easy_cleanup(e);

                    //read one handle from easy handle queue and add using **curl_multi_add_handle.**  (thread-safe)
                } 
         }       
} while (1);
```

**Reply by jay** [reply]
> This is the gdb output of the thread executing multi_perform -
> #0 0x0000003936594684 in PEM_read_bio_ex () from /usr/lib64/libcrypto.so.1.1

It is likely the cpu usage is coming from openssl. As I suggested earlier I would profile it to confirm.

> I did not use CURLMOPT_MAX_TOTAL_CONNECTIONS and CURLMOPT_MAX_HOST_CONNECTIONS, because in case of delay on one message, I did not want to delay the transfer of another, which would have taken place on the same set of connections (if the limit has been set). Is my understanding correct?

It depends on whether both the client (libcurl) and server support HTTP/2 and it can be negotiated. curl will multiplex by default for HTTP/2 connections since 7.62.0 so once you have an HTTP/2 connection established then curl can perform multiple transfers on that connection at the same time _if possible_. Also for HTTP/2 you may want to enable [CURLOPT_PIPEWAIT](https://curl.se/libcurl/c/CURLOPT_PIPEWAIT.html) depending on your circumstance. In other cases curl is using HTTP/1 and must wait for the current transfer on the connection to complete before re-using the connection for another transfer.

_edit: I assumed when I wrote this that the requests are to the same hostname but looking through this discussion I may have been mistaken. Maybe I was thinking about a different discussion. If the requests are to different hostnames then yes there would be separate connections and setting max total connections would limit how many could happen at once._

> Another question, is it necessary to complete all the transfers, before adding any new easy handles to the multi?

No.

**Reply by bansalnehal** [reply]
I am using profiling for the first time. This is what I captured: 

![perf](https://github.com/curl/curl/assets/47805991/b3d8c0c5-bee6-4b95-a685-d4f61dc5850e)

How should I better use it? 

> It depends on whether both the client (libcurl) and server support HTTP/2 and it can be negotiated. curl will multiplex by default for HTTP/2 connections since 7.62.0 so once you have an HTTP/2 connection established then curl can perform multiple transfers on that connection at the same time if possible. Also for HTTP/2 you may want to enable [CURLOPT_PIPEWAIT](https://curl.se/libcurl/c/CURLOPT_PIPEWAIT.html) depending on your circumstance. In other cases curl is using HTTP/1 and must wait for the current transfer on the connection to complete before re-using the connection for another transfer.

The transfer is with the same hostname. 
There were multiple observations regarding multiplexing: 
1) I am using 7.76 with Centos6.10. Beyond 7.62, I read it on the website that multiplexing was enabled by default and pipelining bit was deprecated. Therefore, I did not set the option explicitly. 
2) I have set the easy option to http2 but in my verbose output, it said that the bundle does not support multi-use, and was sending the requests over http1.1. 

Is this in line with what should have happened?

**Reply by jay** [reply, marked as answer]
I think there is definitely an issue here with OpenSSL using CPU to parse the list but the OpenSSL version you are using is no longer supported by them. I think a lot of the problem with this is you are using old versions of libraries and they all have fixed many bugs since then. So whether it's a bug or just some resource limitation I don't know.

> 1. I am using 7.76 with Centos6.10. Beyond 7.62, I read it on the website that multiplexing was enabled by default and pipelining bit was deprecated. Therefore, I did not set the option explicitly.
> 2. I have set the easy option to http2 but in my verbose output, it said that the bundle does not support multi-use, and was sending the requests over http1.1.
> 
> Is this in line with what should have happened?

As I explained both libcurl and the server must support and agree to use http/2. There is no pipelining for HTTP/1, it hasn't been supported for a very long time. Please refer to [CURLMOPT_PIPELINING](https://curl.se/libcurl/c/CURLMOPT_PIPELINING.html).

**Reply by bansalnehal** [reply]
Okay, I'll see with the upgraded versions of both. 

One last question, is it possible to determine **if I have given an unreachable server** or **the transfer has timed out** using error code returned in msg->data.result after curl_multi_info_read(). 

I tried giving an unreachable server but the error code returned is 28, which is the same error code returned when there is a transfer delay. 
I am currently using the connect time and total time returned with easy handle on completed transfer to figure this out.

**Comment by dfandrich** [comment]
A properly-operating multi loop won't take 100% CPU if the server isn't reachable.

**Reply by bansalnehal** [reply]
Actually, the thread 2 that I am using is to execute curl multi perform on the requests that are being added by Thread 1. So there might be times when the function has to execute on empty handle. 
Also, if I use multi wait before multi perform on thread 2, then due to unavailability of lock, thread 1 won’t be able to add more requests. 

Please suggest how to proceed.

**Reply by bansalnehal** [reply]
I changed the implementation as suggested above, and yes when the server is unreachable, the CPU is at 3%.

**Comment by dfandrich** [comment]
An unreachable server should return error 7 while a timeout should return 28.
But, if the timeout is too short to determine that the server is unreachable,
you could get 28 in both cases.

**Reply by bansalnehal** [reply]
Okay. Then that must be happening in my case. Because my timeout is 1 second.

**Comment by bansalnehal** [comment]
Hi @jay @bagder , I encountered a strange scenario with my application. The requests created by the multi are getting timed out without the server receiving them. I checked the netstat -apn output and it showed thousands of connections to the server in TIME_WAIT state and a few in ESTABLISHED state. What could be the reason for this? I have set CONNECTTIMEOUT to 1 sec and TIMEOUT to 2 sec.

**Reply by jay** [reply]
Those are very short timeouts and are surely limiting the transfers' chances of success.

**Reply by bansalnehal** [reply]
Hi, the server is responding in time for keeping the timeout values so short. 
The actual problem is this: 
The number of TCP connections have increased to the order of 5000 and almost all of them are in TIME_WAIT/FIN_WAIT2 with hardly 50 in ESTABLISHED state. Is there a specific reason for this? Also, does curl consider them to be in the connection pool to check for an active connection to send a new request? Now my application is not even sending the requests in time to the server.

**Reply by bansalnehal** [reply]
Hi, I did some debugging by adding prints. Specified a port range of 25000 ports starting from 33000, however whenever the connect time is 0, the port number is 0 for that request. What could be the reason?

**Reply by jay** [reply]
TIME_WAIT is part of the TCP protocol and is the last stage of TCP shutdown before the connection CLOSED final state. TIME_WAIT is to wait some time for any delayed packets. It has already been suggested that you are not giving nearly enough time to allow the transfers to succeed. Also it sounds like you also have resource issues.

**Reply by bansalnehal** [reply]
I have done the following optimizations: 
1) Increased the overall timeout value to 3 seconds with no timer on connection time. 
2) Put a limit on the max_connects and max_total_connections of 1000. 

Even with the above changes, if the server is faulty, requests are getting failed.

The requests are allowed to fail only if there has been an attempt made to connect with the server. Is there a way to figure out if the attempt to make a connection with the server failed at TCP layer itself?.  

Because as per my observations, if the server continues to lag, my application restarts after making 19000  TCP connections (these include ESTABLISHED as well as TIME_WAIT). Why is the max_connects limit not working with this? Also, can I tweak my OS parameters to resolve this?

**Reply by jay** [reply]
> 1. Increased the overall timeout value to 3 seconds with no timer on connection time.

It is important to understand that when you set the maximum time a transfer is allowed that includes queue time. Refer to 
[CURLMOPT_MAX_TOTAL_CONNECTIONS](https://curl.se/libcurl/c/CURLMOPT_MAX_TOTAL_CONNECTIONS.html):

_When more transfers are added to the multi handle than what can be performed due to the set limit, they get queued up waiting for their chance. When that happens, the [CURLOPT_TIMEOUT_MS](https://curl.se/libcurl/c/CURLOPT_TIMEOUT_MS.html) timeout is counted inclusive of the waiting time, meaning that if you set a too narrow timeout in such a case the transfer might never even start before it times out._

> Even with the above changes, if the server is faulty, requests are getting failed.

What error code are you getting for the transfers? CURLE_OPERATION_TIMEDOUT? 

> my application restarts after making 19000 TCP connections. Why is the max_connects limit not working with this?

CURLMOPT_MAX_TOTAL_CONNECTIONS is the maximum amount of simultaneously open connections

**Reply by bansalnehal** [reply]
> When more transfers are added to the multi handle than what can be performed due to the set limit, they get queued up waiting for their chance. When that happens, the [CURLOPT_TIMEOUT_MS](https://curl.se/libcurl/c/CURLOPT_TIMEOUT_MS.html) timeout is counted inclusive of the waiting time, meaning that if you set a too narrow timeout in such a case the transfer might never even start before it times out.

This is my main concern if the requests are getting timed out in the queue itself. I am using curl 7.76 and the queue time is not extractable in this version from the completed requests. How to figure this out in 7.76 version?

> What error code are you getting for the transfers? CURLE_OPERATION_TIMEDOUT?

Yes, mostly this is the error code. Error code 7 is also seen sometimes, though it is very less when compared with CURLE_OPERATION-TIMEDOUT. 

> CURLMOPT_MAX_TOTAL_CONNECTIONS is the maximum amount of simultaneously open connections

I think this is only governing the connections in ESTABLISHED state not the ones in TIME_WAIT. Could it be this?

**Reply by jay** [reply]
The majority of the transfers are timing out due to the limitation you set. [libcurl-errors](https://curl.se/libcurl/c/libcurl-errors.html)

CURLINFO_QUEUE_TIME_T is only available since 8.6.0.

If libcurl calls Curl_disconnect which will likely call sclose (closesocket, etc) then that is not considered an open connection.

**Reply by bansalnehal** [reply]
What could be a possible way out? The timeout is required to be set for the application. 
I removed the timeout completely. But now my application is restarting on reaching 1000 connections as set in MAX_TOTAL_CONNECTIONS.

**Reply by bansalnehal** [reply]
There is an additional observation. When I set the limit to 500 MAX_TOTAL_CONNECTIONS, it is crashing in multi perform with the following stack trace: 
/lib64/libpthread.so.0() [0x37c280f7e0]
/lib64/libpthread.so.0(__pthread_rwlock_rdlock+0xa) [0x37c280aa4a]
/usr/lib64/libcrypto.so.1.1(CRYPTO_THREAD_read_lock+0x9) [0x39365e1d79]
/usr/lib64/libcrypto.so.1.1() [0x3936554f00]
/usr/lib64/libcrypto.so.1.1(ERR_lib_error_string+0x4e) [0x393655507e]
/usr/lib64/libcrypto.so.1.1(ERR_error_string_n+0x84) [0x3936555114]
/usr/lib64/libcurl.so.4(+0x5d2a1) [0x7f865c5b42a1]
/usr/lib64/libcurl.so.4(+0x605ed) [0x7f865c5b75ed]
/usr/lib64/libcurl.so.4(+0x6137e) [0x7f865c5b837e]
/usr/lib64/libcurl.so.4(+0x62f72) [0x7f865c5b9f72]
/usr/lib64/libcurl.so.4(+0x256d6) [0x7f865c57c6d6]
/usr/lib64/libcurl.so.4(+0x2867b) [0x7f865c57f67b]
/usr/lib64/libcurl.so.4(+0x36d95) [0x7f865c58dd95]
/usr/lib64/libcurl.so.4(curl_multi_perform+0x10b) [0x7f865c58eb6b]

Could you tell the possible reason as now I have used multi handle on a single thread. There is no sharing of the handle on multiple threads. And in total there are two threads running with different multi handle on each. Curl version is 7.76

**Reply by jay** [reply]
I reiterate my previous advice to use the latest libcurl and OpenSSL. You might then try ThreadSanitizer.

**Comment by dfandrich** [comment]
If that new issue is not related to this discussion, please open a new issue for
it. It gets too confusing otherwise. Also, read
https://curl.se/libcurl/c/threadsafe.html before you do.
